staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/managedfieldsupdater.go

```diff
@@ -61,15 +64,44 @@ func (f *managedFieldsUpdater) Update(liveObj, newObj runtime.Object, managed Ma
 		} else {
 			managed.Fields()[manager] = vs
 		}
-		// Update the time only if the manager's fieldSet has changed.
-		if formerSet == nil || !managed.Fields()[manager].Set().Equals(formerSet.Set()) {
+
+		currentSet := managed.Fields()[manager]
+		update, err := f.timeShouldBeUpdated(formerSet, currentSet, liveObj, newObj)
+		if err != nil {
+			return object, managed, err
+		}
+		if update {
 			managed.Times()[manager] = &metav1.Time{Time: time.Now().UTC()}
 		}
 	}
 
 	return object, managed, nil
 }
 
+// Update the time upon creation, or if the manager's fieldSet has changed,
+// or if one of the managed field value has changed.
+func (f *managedFieldsUpdater) timeShouldBeUpdated(formerSet, currentSet fieldpath.VersionedSet, liveObj, newObj runtime.Object) (bool, error) {
+	if formerSet == nil {
+		return true, nil
+	}
+	if !currentSet.Set().Equals(formerSet.Set()) {
+		return true, nil
+	}
+
+	liveObjTyped, err := f.typeConverter.ObjectToTyped(liveObj)
+	if err != nil {
+		return false, fmt.Errorf("failed to convert live object to typed: %v", err)
+	}
+	newObjTyped, err := f.typeConverter.ObjectToTyped(newObj)
+	if err != nil {
+		return false, fmt.Errorf("failed to convert new object to typed: %v", err)
+	}
+
+	compare, _ := liveObjTyped.Compare(newObjTyped)
+	modifiedManagedFields := currentSet.Set().Intersection(compare.Modified)
```

Comment:
You shouldn't have to perform a full new comparison, especially since we already have one done above (L52):
```golang
object, managed, err := f.fieldManager.Update(liveObj, newObj, managed, self)
```

We've done a similar thing for the `Apply` use-case, as you can see here: https://github.com/kubernetes-sigs/structured-merge-diff/blob/master/merge/update.go#L207-L210

We should be able to use the same approach. I think the change would be a lot simpler if you did it that way.

Also, this didn't break any existing test? That's really unfortunate.